### PR TITLE
[FLINK-2105] Implement Sort-Merge Outer Join algorithm

### PR DESCRIPTION
**Pull request by r-pogalz**
This pull request relates to [FLINK-2105](https://issues.apache.org/jira/browse/FLINK-2105) and includes:
* Extraction of the MergeMatchIterator logic, so that also the SMOuterJoinIterators can use it
* Extension of the TestData.Generator to provide meaningful test cases
* Implementation of the NonReusing- and ReusingMergeOuterJoinIterator + test cases for both iterators

### CODE REVIEW

File: flink-runtime/src/main/java/org/apache/flink/runtime/operators/sort/AbstractMergeMatchIterator.java [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+import org.apache.flink.util.Collector;
+import org.apache.flink.util.MutableObjectIterator;
+
+import java.util.Iterator;
+
+/**
+ * An implementation of the {@link org.apache.flink.runtime.operators.util.JoinTaskIterator} that realizes the
+ * matching through a sort-merge join strategy.
+ */
+public abstract class AbstractMergeMatchIterator<T1, T2, O> extends AbstractMergeIterator<T1, T2, O> {
```
Comment (fhueske) [inline flink-runtime/src/main/java/org/apache/flink/runtime/operators/sort/AbstractMergeMatchIterator.java]: Rename to AbstractMerge**InnerJoin**Iterator. Match is a term we used back in the old Stratosphere days ;-)

File: flink-runtime/src/main/java/org/apache/flink/runtime/operators/sort/NonReusingMergeMatchIterator.java [hunk 2]
```diff
@@ -18,60 +18,19 @@
...
-	private final NonReusingBlockResettableIterator<T2> blockIt;	// for N:M cross products with same key
-
-	private final List<MemorySegment> memoryForSpillingIterator;
-
-	private final MemoryManager memoryManager;
-
-	private final IOManager ioManager;
-
-	// --------------------------------------------------------------------------------------------
+public class NonReusingMergeMatchIterator<T1, T2, O> extends AbstractMergeMatchIterator<T1, T2, O> {
```
Comment (fhueske) [inline flink-runtime/src/main/java/org/apache/flink/runtime/operators/sort/NonReusingMergeMatchIterator.java]: Rename to NonReusingMerge**InnerJoin**Iterator.

File: flink-runtime/src/main/java/org/apache/flink/runtime/operators/sort/ReusingMergeMatchIterator.java [hunk 3]
```diff
@@ -18,70 +18,20 @@
...
-	private T2 copy2;
-	
-	private T2 blockHeadCopy;
-	
-	private final NonReusingBlockResettableIterator<T2> blockIt;				// for N:M cross products with same key
-	
-	private final List<MemorySegment> memoryForSpillingIterator;
-	
-	private final MemoryManager memoryManager;
+public class ReusingMergeMatchIterator<T1, T2, O> extends AbstractMergeMatchIterator<T1, T2, O> {
```
Comment (fhueske) [inline flink-runtime/src/main/java/org/apache/flink/runtime/operators/sort/ReusingMergeMatchIterator.java]: Rename to ReusingMerge**InnerJoin**Iterator.

**Comment by fhueske** [comment]
Hi @jkovacs and @r-pogalz,
thanks for this PR!

It takes a while to review the changes because you are refactoring parts of the existing code (which is good IMO). Started reviewing the PR but I'm not done yet. I checked the restructuring of the existing MergeJoinIterators and everything is looking good so far. Nice work!
I'll have a look at the new code (MergeOuterJoinIterators) in the next days.

**Comment by chiwanpark** [comment]
Hi, I am reviewing this changes. I'm not done yet but I found some points which are able to improve.

First, there are some duplicated classes such as `SimpleFlatJoinFunction`, `MatchRemovingMatcher`, `Match`, `CollectionIterator`. I think that we can this classes move under `org.apache.flink.runtime.operators.testutils` package. After moving them, they can be shared with test cases for hash-based outer join.

Second, this is just my opinion, how about creating iterator classes for each outer join type such as, `AbstractMergeLeftOuterJoinIterator`, `AbstractMergeRightOuterJoinIterator`, `AbstractMergeFullOuterJoinIterator` and derived classes by reusing variable? I'm concerned about time consuming by comparing outer join type for many records in `callWithNextKey` method. The outer join type is already decided before doing join operation. But I'm not sure that there is obvious performance decrease by this comparing. If the performance decrease is negligible, the second suggestion could be ignored.

**Comment by fhueske** [comment]
+1 for @chiwanpark's first point

I was actually also thinking about separating different join types for performance. What could also help is to make the `outerJoinType` variable `final`. That might help the JIT compiler. Since separating the classes would mean 9 classes instead of 3, I would suggest to implement a micro-benchmark (only for one of the three types) to check the performance implications.

File: flink-runtime/src/main/java/org/apache/flink/runtime/operators/sort/AbstractMergeOuterJoinIterator.java [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+
+/**
+ * An implementation of the {@link org.apache.flink.runtime.operators.util.JoinTaskIterator} that realizes the
+ * outer join through a sort-merge join strategy.
+ * */
+public abstract class AbstractMergeOuterJoinIterator<T1, T2, O> extends AbstractMergeIterator<T1, T2, O> {
+
+	public static enum OuterJoinType {LEFT, RIGHT, FULL}
+
+	private Boolean it1Empty;
```
Comment (fhueske) [inline flink-runtime/src/main/java/org/apache/flink/runtime/operators/sort/AbstractMergeOuterJoinIterator.java]: why aren't you using the {{boolean}} primitive type?

File: flink-runtime/src/main/java/org/apache/flink/runtime/operators/sort/AbstractMergeOuterJoinIterator.java [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+			TypeSerializer<T2> serializer2, TypeComparator<T2> comparator2,
+			TypePairComparator<T1, T2> pairComparator,
+			MemoryManager memoryManager,
+			IOManager ioManager,
+			int numMemoryPages,
+			AbstractInvokable parentTask)
+			throws MemoryAllocationException {
+		super(input1, input2, serializer1, comparator1, serializer2, comparator2, pairComparator, memoryManager, ioManager, numMemoryPages, parentTask);
+
+		this.outerJoinType = outerJoinType;
```
Comment (fhueske) [inline flink-runtime/src/main/java/org/apache/flink/runtime/operators/sort/AbstractMergeOuterJoinIterator.java]: please initialize {{it1Empty}} and {{it2Empty}}.

File: flink-runtime/src/main/java/org/apache/flink/runtime/operators/sort/AbstractMergeOuterJoinIterator.java [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+				it1Empty = true;
+				return false;
+			}
+		}
+		if (it1Empty) {
+			if(outerJoinType == OuterJoinType.RIGHT || outerJoinType == OuterJoinType.FULL) {
+				joinRightKeyValuesWithNull(iterator2.getValues(), matchFunction, collector);
+				it2Empty = !iterator2.nextKey();
+				return true;
+			}else{
```
Comment (fhueske) [inline flink-runtime/src/main/java/org/apache/flink/runtime/operators/sort/AbstractMergeOuterJoinIterator.java]: code style, please add whitespaces: `} else {`

File: flink-runtime/src/main/java/org/apache/flink/runtime/operators/sort/AbstractMergeOuterJoinIterator.java [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+	 * from different inputs. Furthermore, depending on the outer join type (LEFT, RIGHT, FULL), all key-value pairs where no
+	 * matching partner from the other input exists are joined with null.
+	 * The output of the <code>match()</code> method is forwarded.
+	 *
+	 * @throws Exception Forwards all exceptions from the user code and the I/O system.
+	 * @see org.apache.flink.runtime.operators.util.JoinTaskIterator#callWithNextKey(org.apache.flink.api.common.functions.FlatJoinFunction, org.apache.flink.util.Collector)
+	 */
+	@Override
+	public boolean callWithNextKey(final FlatJoinFunction<T1, T2, O> matchFunction, final Collector<O> collector) throws Exception {
+		if(it1Empty == null && it2Empty == null) {
```
Comment (fhueske) [inline flink-runtime/src/main/java/org/apache/flink/runtime/operators/sort/AbstractMergeOuterJoinIterator.java]: You can check the first call with another `boolean` flag. Checking against a non-initialized variable is not so nice, IMO.

File: flink-runtime/src/main/java/org/apache/flink/runtime/operators/sort/AbstractMergeOuterJoinIterator.java [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+ * outer join through a sort-merge join strategy.
+ * */
+public abstract class AbstractMergeOuterJoinIterator<T1, T2, O> extends AbstractMergeIterator<T1, T2, O> {
+
+	public static enum OuterJoinType {LEFT, RIGHT, FULL}
+
+	private Boolean it1Empty;
+	private Boolean it2Empty;
+
+	private OuterJoinType outerJoinType;
```
Comment (fhueske) [inline flink-runtime/src/main/java/org/apache/flink/runtime/operators/sort/AbstractMergeOuterJoinIterator.java]: Please make `outerJoinType` `final`.

File: flink-runtime/src/main/java/org/apache/flink/runtime/operators/sort/AbstractMergeOuterJoinIterator.java [hunk 9]
```diff
@@ -0,0 +1,189 @@
...
+		if(it1Empty == null && it2Empty == null) {
+			//first run, set iterators to first elements
+			it1Empty = !this.iterator1.nextKey();
+			it2Empty = !this.iterator2.nextKey();
+		}
+
+		if (it1Empty && it2Empty) {
+			return false;
+		}
+		if (it2Empty) {
```
Comment (fhueske) [inline flink-runtime/src/main/java/org/apache/flink/runtime/operators/sort/AbstractMergeOuterJoinIterator.java]: Please use `else if` to make it more obvious that these are conditions are mutually exclusive.

File: flink-runtime/src/main/java/org/apache/flink/runtime/operators/sort/AbstractMergeOuterJoinIterator.java [hunk 10]
```diff
@@ -0,0 +1,189 @@
...
+				return true;
+			}else{
+				//consume rest of right side
+				while(iterator2.nextKey());
+				it2Empty = true;
+				return false;
+			}
+		}
+
+		final TypePairComparator<T1, T2> comparator = super.pairComparator;
```
Comment (fhueske) [inline flink-runtime/src/main/java/org/apache/flink/runtime/operators/sort/AbstractMergeOuterJoinIterator.java]: Please put this in an `else` block.

File: flink-runtime/src/test/java/org/apache/flink/runtime/operators/sort/NonReusingSortMergeOuterJoinIteratorITCase.java [hunk 11]
```diff
@@ -0,0 +1,593 @@
...
+																		  OuterJoinType outerJoinType) throws Exception {
+		input1.reset();
+		input2.reset();
+		NonReusingMergeOuterJoinIterator<Tuple2<String, String>, Tuple2<String, Integer>, Tuple4<String, String, String, Object>> iterator =
+				new NonReusingMergeOuterJoinIterator<Tuple2<String, String>, Tuple2<String, Integer>, Tuple4<String, String, String, Object>>(
+						outerJoinType, input1, input2, serializer1, comparator1, serializer2, comparator2,
+						pairComp, this.memoryManager, this.ioManager, PAGES_FOR_BNLJN, this.parentTask);
+
+		List<Tuple4<String, String, String, Object>> actual = new ArrayList<Tuple4<String, String, String, Object>>();
+		while (iterator.callWithNextKey(new SimpleFlatJoinFunction(), new ListCollector<Tuple4<String, String, String, Object>>(actual))) ;
```
Comment (fhueske) [inline flink-runtime/src/test/java/org/apache/flink/runtime/operators/sort/NonReusingSortMergeOuterJoinIteratorITCase.java]: I think you don't need to create a new ListCollector() in each while iteration.

File: flink-runtime/src/test/java/org/apache/flink/runtime/operators/sort/ReusingSortMergeOuterJoinIteratorITCase.java [hunk 12]
```diff
@@ -0,0 +1,587 @@
...
+
+	// --------------------------------------------------------------------------------------------
+	//                                    Utilities
+	// --------------------------------------------------------------------------------------------
+
+
+	/**
+	 * custom function that simply joins two tuples and considers null cases
+	 */
+	private static class SimpleFlatJoinFunction implements FlatJoinFunction<Tuple2<String, String>, Tuple2<String, Integer>, Tuple4<String, String, String, Object>> {
```
Comment (fhueske) [inline flink-runtime/src/test/java/org/apache/flink/runtime/operators/sort/ReusingSortMergeOuterJoinIteratorITCase.java]: The utilities of the `ReusingSortMergeOuterJoinIteratorITCase` and `NonReusingSortMergeOuterJoinIteratorITCase` look very similar (didn't check in detail). Can you put them in an abstract class that both test classes extend?

**Comment by fhueske** [comment]
Very nice work! :+1: I had just a few minor style regards.

It would be nice if you could check the overhead of having all outer join cases handled by the same class vs. having a dedicated class for LEFT, RIGHT, and FULL OUTER. 

I would also like to run a major regression benchmark to double-check that the refactoring of the current sort-merge join didn't cause a performance regression. I can take care of that once you addressed the comments.

If all lights are green after that, I would be happy to merge this PR.

**Comment by jkovacs** [comment]
Thanks @fhueske and @chiwanpark for the comprehensive review. I have pushed our updated branch, addressing your comments. @r-pogalz will also contribute a microbenchmark in the next couple of days.

File: flink-runtime/src/main/java/org/apache/flink/runtime/operators/sort/AbstractMergeInnerJoinIterator.java [hunk 13]
```diff
@@ -0,0 +1,108 @@
...
+			MemoryManager memoryManager,
+			IOManager ioManager,
+			int numMemoryPages,
+			AbstractInvokable parentTask)
+			throws MemoryAllocationException {
+		super(input1, input2, serializer1, comparator1, serializer2, comparator2, pairComparator, memoryManager, ioManager, numMemoryPages, parentTask);
+	}
+
+	/**
+	 * Calls the <code>JoinFunction#match()</code> method for all two key-value pairs that share the same key and come
```
Comment (fhueske) [inline flink-runtime/src/main/java/org/apache/flink/runtime/operators/sort/AbstractMergeInnerJoinIterator.java]: `match()` function has been renamed to `join()`. Please update, thanks!

File: flink-runtime/src/main/java/org/apache/flink/runtime/operators/sort/AbstractMergeInnerJoinIterator.java [hunk 14]
```diff
@@ -0,0 +1,108 @@
...
+	 * from different inputs. The output of the <code>match()</code> method is forwarded.
+	 * <p/>
+	 * This method first zig-zags between the two sorted inputs in order to find a common
+	 * key, and then calls the match stub with the cross product of the values.
+	 *
+	 * @throws Exception Forwards all exceptions from the user code and the I/O system.
+	 * @see org.apache.flink.runtime.operators.util.JoinTaskIterator#callWithNextKey(org.apache.flink.api.common.functions.FlatJoinFunction, org.apache.flink.util.Collector)
+	 */
+	@Override
+	public boolean callWithNextKey(final FlatJoinFunction<T1, T2, O> matchFunction, final Collector<O> collector)
```
Comment (fhueske) [inline flink-runtime/src/main/java/org/apache/flink/runtime/operators/sort/AbstractMergeInnerJoinIterator.java]: Please rename parameter to `joinFunction`.

File: flink-runtime/src/main/java/org/apache/flink/runtime/operators/sort/AbstractMergeInnerJoinIterator.java [hunk 15]
```diff
@@ -0,0 +1,108 @@
...
+				current2 = this.iterator2.getCurrent();
+			} else {
+				if (!this.iterator1.nextKey()) {
+					return false;
+				}
+				comparator.setReference(this.iterator1.getCurrent());
+			}
+		}
+
+		// here, we have a common key! call the match function with the cross product of the
```
Comment (fhueske) [inline flink-runtime/src/main/java/org/apache/flink/runtime/operators/sort/AbstractMergeInnerJoinIterator.java]: match -> join

File: flink-runtime/src/main/java/org/apache/flink/runtime/operators/sort/AbstractMergeInnerJoinIterator.java [hunk 16]
```diff
@@ -0,0 +1,108 @@
...
+		// values
+		final Iterator<T1> values1 = this.iterator1.getValues();
+		final Iterator<T2> values2 = this.iterator2.getValues();
+
+		crossMatchingGroup(values1, values2, matchFunction, collector);
```
Comment (fhueske) [inline flink-runtime/src/main/java/org/apache/flink/runtime/operators/sort/AbstractMergeInnerJoinIterator.java]: matchFunction -> joinFunction

File: flink-runtime/src/main/java/org/apache/flink/runtime/operators/sort/AbstractMergeIterator.java [hunk 17]
```diff
@@ -138,63 +122,23 @@ public void abort() {
 	 * key, and then calls the match stub with the cross product of the values.
```
Comment (fhueske) [inline flink-runtime/src/main/java/org/apache/flink/runtime/operators/sort/AbstractMergeIterator.java]: match() -> join()
match stub -> join function

File: flink-runtime/src/main/java/org/apache/flink/runtime/operators/sort/AbstractMergeIterator.java [hunk 18]
```diff
@@ -138,63 +122,23 @@ public void abort() {
...
-	throws Exception
-	{
-		if (!this.iterator1.nextKey() || !this.iterator2.nextKey()) {
-			// consume all remaining keys (hack to prevent remaining inputs during iterations, lets get rid of this soon)
-			while (this.iterator1.nextKey());
-			while (this.iterator2.nextKey());
-			
-			return false;
-		}
+	public abstract boolean callWithNextKey(final FlatJoinFunction<T1, T2, O> matchFunction, final Collector<O> collector)
```
Comment (fhueske) [inline flink-runtime/src/main/java/org/apache/flink/runtime/operators/sort/AbstractMergeIterator.java]: matchFunction -> joinFunction

File: flink-runtime/src/main/java/org/apache/flink/runtime/operators/sort/AbstractMergeIterator.java [hunk 19]
```diff
@@ -138,63 +122,23 @@ public void abort() {
...
-				comparator.setReference(this.iterator1.getCurrent());
-			}
-		}
-		
-		// here, we have a common key! call the match function with the cross product of the
-		// values
-		final NonReusingKeyGroupedIterator<T1>.ValuesIterator values1 = this.iterator1.getValues();
-		final NonReusingKeyGroupedIterator<T2>.ValuesIterator values2 = this.iterator2.getValues();
-		
+	protected void crossMatchingGroup(Iterator<T1> values1, Iterator<T2> values2, FlatJoinFunction<T1, T2, O> matchFunction, Collector<O> collector) throws Exception {
```
Comment (fhueske) [inline flink-runtime/src/main/java/org/apache/flink/runtime/operators/sort/AbstractMergeIterator.java]: matchFunction -> joinFunction

**Comment by fhueske** [comment]
Thanks for the update @r-pogalz and @jkovacs!

Looks really good. I would like to ask you for one more favor. As you might know, Join was previously called Match and the name was changed quite some time ago. All public interfaces were updated but the name Match lives on in many places in the internals. It would be great if you could update `match` to `join` in the classes that you rewrite (esp. in JavaDocs and parameter names). I started adding comments for that, but stopped at one point.

@r-pogalz looking forward to the microbenchmark!

Thank you, Fabian

**Comment by r-pogalz** [comment]
@fhueske, I ran through the classes and tried to catch all occurrences of `match` and replaced them by `join`. Hopefully I did not miss something :)

Moreover, I set up a benchmark with JMH and did several runs on my machine (MacBook Air, 1.3 GHz Intel Core i5, 4 GB 1600 MHz DDR3). I pushed the code on another branch, so you can check it for further information (commit can be seen [here](https://github.com/r-pogalz/flink/commit/629c2ff5caa9775d968deecdc2ff0b965310a09e)). I decided to test a dedicated class for the Left Outer Join against the left outer join using the existing class for all types. An iteration in the benchmark measures the time needed to perform the outer join for the entire inputs. For the first couple of runs I chose a fixed cardinality of 10k entries for the right input. I varied the input size of the left input (20k, 50k and 100k).

 Cardinality of left input | Same class (avg. time in seconds) | Dedicated class (avg. time in seconds)
------------ | ------------- | ------------
20k | **4.264** | 4.282
50k | **7.845** | 7.913
100k | 14.090 | **13.907**

Moreover, I did a test run with **50k entries on both inputs** and got the following results:

Same class (avg. time in seconds) | Dedicated class (avg. time in seconds)
-------------- | ---------------
**10.251** | 10.571

The results show that both implementations perform almost similar. So I guess the JIT compiler does proper optimizations concerning the if-statements in our current implementation. Or what do you think @fhueske and @chiwanpark ?

I also pushed the results. Please look [here](https://github.com/r-pogalz/flink/commit/32dfabe33efc382ee2bb1c879eba507318cdbc24) for detailed information (e.g. min, max, std. deviation).

**Comment by StephanEwen** [comment]
For most other tests that we have added/reworked in the past, we moved away from using `Record` for the tests. The `Record` was the core element of the old record API, which is deprecated.

Using simple Java types makes the tests usually easier, more readable, and closer to the real behavior of the system (with respect to object reuse and mutability).

Would be nice to move the tests from `Record` to `Tuple2<Long, String>` or something comparable.

**Comment by fhueske** [comment]
Hi @r-pogalz & @jkovacs,
Thanks for doing the microbenchmarks, posting the performance numbers, and moving the tests to `Tuple2<Integer, String>`.

I think this PR is good to merge.

**Comment by r-pogalz** [comment]
Hello @fhueske,
I see you noticed that we replaced all occurrences of `Record` by `Tuple2<Integer, String>` in the InnerJoin- and OuterJoinITCases. The issue is that we did not yet adapt the other tests (~15 test classes) using the `TestData.Generator`. That's why we currently added a second Generator (`TestData.TupleGenerator`) for our purposes without the usage of `Record`. But I think it would be good to replace the old `TestData.Generator` by the new one and adapt all usages at some point in the future.

**Comment by fhueske** [comment]
Yes, I noticed that, thanks for pointing it out. :-)
I think it is fine this way. The PR is already quite big and we shouldn't put to much things in.
It would be good, if you could open a JIRA to make sure we remember updating the remaining tests.
Thanks!

**Comment by r-pogalz** [comment]
Alright, I created a ticket ([FLINK-2479](https://issues.apache.org/jira/browse/FLINK-2479)) for adjustment of the other tests. 

Is there still anything left to do for this PR from our side?

**Comment by fhueske** [comment]
No, I don't think so. :-)
I'll have one more close look and will merge it if everything is OK.

Thanks a lot for the PR and all adoptions!

**Comment by fhueske** [comment]
Merging

**Comment by fhueske** [comment]
Thanks for the excellent work! 
The PR is merged :-)

**Pull request closed without merging**